### ISSUE TITLE
React components?

### ISSUE THREAD

**Issue by rauschma** [choojs/hyperx#9]
How do you handle React components (=function calls)? You could either have a registry for component names or do this:

``` js
class EchoComponent {
    ···
}
let comp = jsx`<${EchoComponent} />`;
React.renderComponent(comp, document.body);
```

**Comment by jgoux** [comment]
I'm interested too. I'd love to use this library but I don't know how to handle custom components.

**Comment by ngoldman** [comment]
Just write a function that accepts parameters and returns a vdom element.

```js
var vdom = require('virtual-dom')
var hyperx=require('hyperx')(vdom.h)

var initState = {
}

function App (state, dispatch) {
  
}

function Component (state, dispatch) {
  var data = { wow: 'cool' }
  return hx`
    <div onclick=${() => dispatch('event', data)}>
      ${state.object.prop}
    </div>
  `
}
```

**Comment by bmathews** [comment]
I've found that you can components using `React.createElement` directly:

```
hx`<button onClick=${this._handleVolumeIconClick}>
  ${React.createElement(Icon, { icon: 'volume-icon })}
</button>
```

Not sure if there's an easier way or not.

**Comment by niksy** [comment]
For anyone looking for a "workaround" for this, I’ve managed to create a proof of concept. Basically, it uses function which returns another function in which we check first letter of tag element. If it’s uppercase, we treat it as component and evaluate that string to variable declaration.

This obviously only works if method is in scope and eval has access to defined variable. Maybe it could be made to work regardless of scope? It seems like this could be properly processed with [hyperxify](https://github.com/substack/hyperxify) - if first letter is uppercased, don’t stringify tag name and just return original value.

```js
function r ( h ) {
	return function ( tagName, attrs, children ) {
		if ( tagName[0].toLowerCase() !== tagName[0] ) {
			tagName = eval(tagName);
		}
		return h(tagName, attrs, children);
	}
}

const React = require('react');
const ReactDOM = require('react-dom');
const hyperx = require('hyperx');
const hx = hyperx(r(React.createElement));
const Slider = require('react-slider');

module.exports = React.createClass({
	render() {
		return hx`
			<div className='controls__toolbar__volume__slider'>
				<Slider value="${this.props.volume} withBars="${true}" />
			</div>`
		;
	}
});
```

Or example of a more complex implementation:

```js
function r ( h ) {
	return function ( tagName, attrs, children ) {
		if ( tagName[0].toLowerCase() !== tagName[0] ) {
			tagName = eval(tagName);
		}
		return h(tagName, attrs, children);
	}
}

const React = require('react');
const ReactDOM = require('react-dom');
const hyperx = require('hyperx');
const hx = hyperx(r(React.createElement));

const Wrap = React.createClass({
	render: function () {
		return hx`<b>${this.props.children}</b>`;
	}
});

const Input = ( props ) => {
	return hx`<input type="text" defaultValue="${props.value}" class="${props.value}" onChange=${props.onChange} />`;
};

const App = React.createClass({
	getInitialState: function () {
		return {
			text: ''
		};
	},
	render: function () {
		return hx`
			<div>
				<span class="bar">
					<Wrap>${['foo','bar'].map(( value, index ) => {
						return hx`<Input value=${value} key="${index}" onChange=${this.handleChange} />`;
					})}
					</Wrap>
				</span>
				<b>${this.state.text}</b>
			</div>
		`;
	},
	handleChange: function ( ev ) {
		this.setState({
			text: ev.target.value
		});
	}
});
  
ReactDOM.render(React.createElement(App), document.querySelector('body'));
```

**Comment by tbranyen** [comment]
@rauschma inspired by your comment: https://github.com/tbranyen/diffhtml/tree/master/packages/diffhtml-components#examples

**Comment by goto-bus-stop** [comment]
as of this PR https://github.com/choojs/hyperx/pull/68, published in 2.4.0, you can use `<${CustomComponent} prop=x> child </${CustomComponent}>`.
